zebra/zebra_mlag_private.c

```diff
@@ -0,0 +1,304 @@
+ * under the terms of the GNU General Public License as published by the Free
+ * Software Foundation; either version 2 of the License, or (at your option)
+ * any later version.
+ *
+ * This program is distributed in the hope that it will be useful, but WITHOUT
+ * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
+ * FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License for
+ * more details.
+ *
+ * You should have received a copy of the GNU General Public License along
+ * with this program; see the file COPYING; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA
+ */
+#include "zebra.h"
+
+#include "hook.h"
+#include "module.h"
+#include "thread.h"
+#include "libfrr.h"
+#include "version.h"
+#include "network.h"
+
+#include "lib/stream.h"
+
+#include "zebra/debug.h"
+#include "zebra/zebra_router.h"
+#include "zebra/zebra_mlag.h"
+#include "zebra/zebra_mlag_private.h"
+
+#include <sys/un.h>
+
+
+/*
+ * This file will have platform specific apis to communicate with MCLAG.
+ *
+ */
+
+#ifdef HAVE_CUMULUS
+
+static struct thread_master *zmlag_master;
+static int mlag_socket;
+
+static int zebra_mlag_connect(struct thread *thread);
+static int zebra_mlag_read(struct thread *thread);
+
+/*
+ * Write the data to MLAGD
+ */
+int zebra_mlag_private_write_data(uint8_t *data, uint32_t len)
+{
+	int rc = 0;
+
+	if (IS_ZEBRA_DEBUG_MLAG) {
+		zlog_debug("%s: Writing %d length Data to clag", __func__, len);
+		zlog_hexdump(data, len);
+	}
+	rc = write(mlag_socket, data, len);
+	return rc;
+}
+
+static void zebra_mlag_sched_read(void)
+{
+	pthread_mutex_lock(&zrouter.mlag_info.mlag_th_mtx);
+	thread_add_read(zmlag_master, zebra_mlag_read, NULL, mlag_socket,
+			&zrouter.mlag_info.t_read);
+	pthread_mutex_unlock(&zrouter.mlag_info.mlag_th_mtx);
+}
+
+static int zebra_mlag_read(struct thread *thread)
+{
+	uint32_t *msglen;
+	uint32_t h_msglen;
+	uint32_t tot_len, curr_len = mlag_rd_buf_offset;
+
+	pthread_mutex_lock(&zrouter.mlag_info.mlag_th_mtx);
+	zrouter.mlag_info.t_read = NULL;
+	pthread_mutex_unlock(&zrouter.mlag_info.mlag_th_mtx);
+
+	/*
+	 * Received message in sock_stream looks like below
+	 * | len-1 (4 Bytes) | payload-1 (len-1) |
+	 *   len-2 (4 Bytes) | payload-2 (len-2) | ..
+	 *
+	 * Idea is read one message completely, then process, until message is
+	 * read completely, keep on reading from the socket
+	 */
+	if (curr_len < ZEBRA_MLAG_LEN_SIZE) {
+		ssize_t data_len;
+
+		data_len = read(mlag_socket, mlag_rd_buffer + curr_len,
+				ZEBRA_MLAG_LEN_SIZE - curr_len);
+		if (data_len == 0 || data_len == -1) {
+			if (IS_ZEBRA_DEBUG_MLAG)
+				zlog_debug("MLAG connection closed socket : %d",
+					   mlag_socket);
+			close(mlag_socket);
+			zebra_mlag_handle_process_state(MLAG_DOWN);
+			return -1;
+		}
+		if (data_len != (ssize_t)ZEBRA_MLAG_LEN_SIZE - curr_len) {
+			/* Try again later */
+			zebra_mlag_sched_read();
+			return 0;
```

Comment:
Initially, when "zebra_mlag_read" is scheduled, curr_len is set to "0", because of read_buffer_reset.
 uint32_t tot_len, curr_len = mlag_rd_buf_offset;

Here Idea is read the socket in 2 steps.
First Header to read the Length of payload. at this time we are updating curre_len with payload length.
curr_len = ZEBRA_MLAG_LEN_SIZE;

after this we read payload. once we read the complete pay_load, processing will be scheduled.

after the processing we are calling Read_buffer_reset, which will reset curr_len = 0

Reply:
Got your concern, Taken care